### ISSUE TITLE
middleware breaks behavior of background task

### ISSUE THREAD

**Issue by jkh911208** [tiangolo/fastapi#2215]
### First check

* [x ] I added a very descriptive title to this issue.
* [ x] I used the GitHub search to find a similar issue and didn't find it.
* [ x] I searched the FastAPI documentation, with the integrated search.
* [ x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [ x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [ x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [ x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [ x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [ x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

```Python
import asyncio
import logging
import time

from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse
from starlette.background import BackgroundTasks

app = FastAPI()


@app.get("/")
async def read_root():
    response_object = {"Hello": "World"}
    tasks = BackgroundTasks()
    tasks.add_task(bg_task)
    return JSONResponse(response_object, 202, background=tasks)


async def bg_task():
    logging.error("test1")
    await asyncio.sleep(10)
    logging.error("test2")


@app.middleware("http")
async def add_process_time_header(request: Request, call_next):
    start_time = time.time()
    response = await call_next(request)
    response.headers["X-Process-Time"] = str(time.time() - start_time)
    response.headers["address"] = request.client.host
    return response

if __name__ == "__main__":
    import uvicorn
    uvicorn.run("main:app", host="0.0.0.0", port=8000,
                log_level="debug", reload=True)

```

### Description

run the above code, make a request to "/"
you will see 
<img width="520" alt="Screen Shot 2020-10-21 at 8 50 42 AM" src="https://user-images.githubusercontent.com/17028674/96729219-89934500-137a-11eb-93a6-70e0bdd49ece.png">
which logging is happening before return the object

but if you make only one request at a time this is fine, but if you make multiple request while background task is running

problem #1 the second request start after first background task finishes
problem #2 when background task of second request finishes, it throw error
<img width="993" alt="Screen Shot 2020-10-21 at 8 52 39 AM" src="https://user-images.githubusercontent.com/17028674/96729600-f1499000-137a-11eb-8cdb-5e4d66c11c5c.png">

### Environment

* OS: [e.g. Linux / Windows / macOS]: macos and debian
* FastAPI Version [e.g. 0.3.0]: 0.61.1
* Python version: 3.8.5

### Additional context

if i remove middleware code it works as expected
<img width="592" alt="Screen Shot 2020-10-21 at 8 55 24 AM" src="https://user-images.githubusercontent.com/17028674/96729820-2d7cf080-137b-11eb-961f-fc655a448073.png">
all requests are handled and all background tasks are working fine no exception

**Comment by ycd** [comment]
I believe this is related to Starlette Issue 919:  [Background tasks don't work with middleware that subclasses `BaseHTTPMiddleware`](https://github.com/encode/starlette/issues/919)

**Comment by Kludex** [comment]
@jkh911208 Could you close this issue, please?

**Comment by yingyingPengSAP** [comment]
For those who are searching for how to call background tasks under FastAPI middleware, here is quick solution: 

from starlette.background import BackgroundTask
from somewhere import functionA

@app.middleware("http")
async def middleware(request: Request, call_next):
        ........(do something)
        background_tasks = BackgroundTask(functionA, args)
        response = await background_tasks.__call__()
    return response

Hope this helps someone!

**Comment by santigandolfo** [comment]
Hi, why is this issue `Closed`? Cause I'm trying to use background tasks and a middleware at the same time, and if I send a request to the endpoint through Postman, I receive my response but the request time doesn't finish until the background task does. But if I remove the middleware then it works as expected.

**Comment by Kludex** [comment]
What is the version of your FastAPI?

**Comment by santigandolfo** [comment]
> What is the version of your FastAPI?
> 
> Oh sorry, yeah, this is still a thing. [encode/starlette#1441](https://github.com/encode/starlette/pull/1441)

I'm on v0.75.2
Do you suggest a workaround until that PR is merged in Starlette and FastAPI bumps it's requirements?

**Comment by Kludex** [comment]
Yes, don't use `BaseHTTPMiddleware`.

**Comment by santigandolfo** [comment]
> Yes, don't use `BaseHTTPMiddleware`.

In that case I couldn't add general before and after request actions, right?

**Comment by Kludex** [comment]
No, I'm saying to not use that class. You can have middlewares: https://pgjones.dev/blog/how-to-write-asgi-middleware-2021
